### DISCUSSION TITLE
Welcome to CBBA-Python Discussions!

### DISCUSSION THREAD

**Discussion by zehuilu** [zehuilu/CBBA-Python#1] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by lylia711** [comment]
I'd like to implement this algorithm to a heterogeneous team of robots and I'd like to know is that possible ?

**Reply by zehuilu** [reply]
Yes, it is possible. The main function of CBBA tasks a list of dataclass **Agent** as input (see [here](https://github.com/zehuilu/CBBA-Python/blob/main/lib/CBBA.py#L115)). For each Agent, you could have different properties like average velocity (see [here](https://github.com/zehuilu/CBBA-Python/blob/main/lib/Agent.py)). Also, you can have multiple types of agents and tasks, where Type-A agents can only execute Type-A tasks. You can define them by _task_type_ of dataclass **Task** and _agent_type_ of dataclass **Agent**. An example is shown [here](https://github.com/zehuilu/CBBA-Python/blob/main/lib/HelperLibrary.py#L20-#L101). The heterogeneity in this repo is the same as what proposed in CBBA paper.

**Comment by xixi87** [comment]
Thank you for your sharing!
I'd like to implement CBBA to ship agents, could you share the closest related article of implement this algorithm to UAV?
I want to understand the codes based on the article.

**Reply by zehuilu** [reply]
Hi xixi87,

Whether agents are aerial robots or ground robots, they shouldn't affect how this algorithm work. You could look into their original paper [https://ieeexplore.ieee.org/abstract/document/5072249](https://ieeexplore.ieee.org/abstract/document/5072249)
As for the application into UAV, here is another paper from them. [https://arc.aiaa.org/doi/abs/10.2514/6.2009-5776](https://arc.aiaa.org/doi/abs/10.2514/6.2009-5776)

**Reply by xixi87** [reply]
Thank you very much!

**Comment by menaxijbagchi** [comment]
Hi,
Can you elaborate a bit about the convergence criteria, i.e., how to know whether the convergence has reached and each agent has the global bid value of a task.

**Reply by zehuilu** [reply]
Hi, I would recommend checking their original paper [Consensus-Based Decentralized Auctions for Robust Task Allocation
](https://ieeexplore.ieee.org/abstract/document/5072249). Particularly, they discuss this in Section V. For a static and connected network, Theorem 1 states that the convergence time is bounded by N_{min}D.

**Comment by PLZCALLMEDIAN** [comment]
Hi ,How to modify the CBBA algorithm with task coupling?
